File: srcr/src/days/day5.rs
use std::collections::HashMap;
use std::fs;
use std::time::Instant;

pub fn solve_day_5() {
    let input = fs::read_to_string("assets/day5.txt").expect("Fichier introuvable");
    let start = Instant::now();

    let (rules, updates) = parse_input(&input);

    let graph = build_graph(&rules);

    let mut sum_of_middles = 0;

    for update in updates {
        let sorted_update = topological_sort(&graph, &update);
        if !sorted_update.is_empty() {
            let middle = sorted_update[sorted_update.len() / 2];
            sum_of_middles += middle;
        }
    }

    println!("Somme des milieux : {}", sum_of_middles);
    println!("Temps total : {:.2?}", start.elapsed());
}

fn parse_input(input: &str) -> (Vec<(i32, i32)>, Vec<Vec<i32>>) {
    let mut lines = input.lines();
    let mut rules = Vec::new();
    let mut updates = Vec::new();

    for line in lines.by_ref() {
        if line.trim().is_empty() {
            break;
        }
        let parts: Vec<i32> = line
            .split('|')
            .map(|x| x.trim().parse::<i32>().unwrap())
            .collect();
        rules.push((parts[0], parts[1]));
    }

    for line in lines {
        let update: Vec<i32> = line
            .split(',')
            .map(|x| x.trim().parse::<i32>().unwrap())
            .collect();
        updates.push(update);
    }

    (rules, updates)
}

fn build_graph(rules: &Vec<(i32, i32)>) -> HashMap<i32, Vec<i32>> {
    let mut graph: HashMap<i32, Vec<i32>> = HashMap::with_capacity(rules.len());
    for &(a, b) in rules {
        graph.entry(a).or_insert(Vec::new()).push(b);
    }
    graph
}

fn topological_sort(graph: &HashMap<i32, Vec<i32>>, update: &Vec<i32>) -> Vec<i32> {
    let mut in_degree: HashMap<i32, usize> = HashMap::new();
    for &node in update {
        in_degree.insert(node, 0);
    }

    for (&node, neighbors) in graph.iter() {
        for &neighbor in neighbors {
            if update.contains(&neighbor) {
                *in_degree.entry(neighbor).or_insert(0) += 1;
            }
        }
    }

    let mut queue: Vec<i32> = Vec::new();
    for &node in update {
        if *in_degree.get(&node).unwrap_or(&0) == 0 {
            queue.push(node);
        }
    }

    let mut result = Vec::new();
    while let Some(current) = queue.pop() {
        result.push(current);
        if let Some(neighbors) = graph.get(&current) {
            for &neighbor in neighbors {
                if update.contains(&neighbor) {
                    let entry = in_degree.entry(neighbor).or_insert(0);
                    *entry -= 1;
                    if *entry == 0 {
                        queue.push(neighbor);
                    }
                }
            }
        }
    }

    result
}

File: srcp/day-5.py
import time
from collections import defaultdict
from srcp.utils.input import get_data

start = time.time()

datas_day_5 = get_data(2024, 5).strip().splitlines()

separator_index = datas_day_5.index("")
rules: list[tuple[int, ...]] = [tuple(map(int, el.split("|"))) for el in datas_day_5[:separator_index]]
updates: list[list[int]] = [list(map(int, el.split(","))) for el in datas_day_5[separator_index + 1:]]

struct_time = time.time() - start


def get_center_page(u: list[int]) -> int:
    return u[len(u) // 2]


def is_update_ordered(tmp_update: list[int], list_rules: list[tuple[int, ...]]) -> bool:
    for r in list_rules:
        if r[0] in tmp_update and r[1] in tmp_update:
            if tmp_update.index(r[0]) > tmp_update.index(r[1]):
                return False
    return True


total: int = 0
for update in updates:
    if is_update_ordered(update, rules):
        total += get_center_page(update)

print(f"Total part 1 : {total}")
end = time.time()
print(f"Time : {end - start:.2f}s")

total_2: int = 0


def build_graph(tmp_rulers: list[tuple[int, ...]]) -> dict[int, list[int]]:
    tmp_graph: dict[int, list[int]] = defaultdict(list)
    for a, b in tmp_rulers:
        tmp_graph[a].append(b)
    return tmp_graph


def topological_sort(tmp_graph: dict[int, list[int]], update: list[int]) -> list[int]:
    in_degree = {u: 0 for u in update}
    for u in update:
        if u in tmp_graph:
            for v in graph[u]:
                if v in update:
                    in_degree[v] += 1

    result = []
    queue = [n for n in update if in_degree[n] == 0]

    while queue:
        tmp = queue.pop(0)
        result.append(tmp)
        if tmp in tmp_graph:
            for v in tmp_graph[tmp]:
                if v in update:
                    in_degree[v] -= 1
                    if in_degree[v] == 0:
                        queue.append(v)

    return result


graph = build_graph(rules)

for update in updates:
    if not is_update_ordered(update, rules):
        sorted_update = topological_sort(graph, update)
        total_2 += get_center_page(sorted_update)

print(f"Total part 2 : {total_2}")
end_2 = time.time()
print(f"Time : {(end_2 - end) + struct_time:.2f}s")
